node/pkg/watchers/evm/watcher.go

```diff
@@ -745,6 +845,65 @@ func canRetryGetBlockTime(err error) bool {
 	return exists
 }
 
+// verifyAndPublish validates a MessagePublication to ensure that it's safe. If so, it broadcasts the message. This function
+// should be the only location where the watcher's msgC channel is written to.
+// Modifies the verificationState field of the message as a side-effect.
+// Even if an invalid Transfer is detected, the message will still be published. It is the responsibility of the calling code to handle
+// a status of Rejected.
+// Note that the result of verification is not returned by this function, but can be accessed directly via the reference to message.
+func (w *Watcher) verifyAndPublish(
+	// Must be non-nil and have verificationState equal to NotVerified.
+	msg *common.MessagePublication,
+	ctx context.Context,
+	// TODO: in practice it might be possible to read the txHash from the MessagePublication and so this argument might be redundant
+	txHash eth_common.Hash,
+	// This argument is only used when Transfer Verifier is enabled. If nil, transfer verifier will fetch the receipt.
+	// Otherwise, the receipt in the calling context can be passed here to save on RPC requests and parsing.
+	receipt *gethTypes.Receipt,
+) error {
+	if msg == nil {
+		return errors.New("verifyAndPublish: message publication cannot be nil")
+	}
+	if msg.VerificationState() != common.NotVerified {
+		return errors.New("verifyAndPublish: message publication already has a verification status")
+	}
+
+	if w.txVerifierEnabled {
+		// This should have already been initialized by the constructor.
+		if w.txVerifier == nil {
+			return errors.New("verifyAndPublish: transfer verifier should be instantiated but is nil")
+		}
+
+		// Setting a custom verification state is only required when
+		// Transfer Verification is enabled. If disabled, the message
+		// will be emitted with the default value `NotVerified`.
+		var verificationState common.VerificationState
+
+		// Verify the transfer by analyzing the transaction receipt. This is a defense-in-depth mechanism
+		// to protect against fraudulent message emissions.
+		valid := w.txVerifier.ProcessEvent(ctx, txHash, receipt)
+		if !valid {
+			w.logger.Error("verifyAndPublish: transfer verification failed", zap.String("txHash", txHash.String()))
+			verificationState = common.Rejected
```

Comment:
Why don't we check for the `tokenBridge` on the VAA before doing the verification here? This way, it would prevent a lot of unnecessary computations.

Reply:
I agree. Not only computation but RPC calls.

Reply:
Should there be a Prometheus metric for verification failures?

Reply:
Done in the latest commit, thanks for catching